### ISSUE TITLE
Rotation rot90 for TTA

### ISSUE THREAD

**Issue by mrkolarik** [ultralytics/yolov5#4107]
## ❔Question
Hi Glenn, 

I am working on a drone image object detection system and I would like to implement custom TTA into our pipeline. I have a general question regarding the TTA implementation (I did read the Test-Time Augmentation (TTA) Tutorial) and possible proposal for a new feature - rot90 TTA augmentation. Here is the original Yolov5 code for TTA augmentation:

``` python 
def forward_augment(self, x):
        img_size = x.shape[-2:]  # height, width
        s = [1, 0.83, 0.67]  # scales
        f = [None, 3, None]  # flips (2-ud, 3-lr)
        y = []  # outputs
        for si, fi in zip(s, f):
            xi = scale_img(x.flip(fi) if fi else x, si, gs=int(self.stride.max()))
            yi = self.forward_once(xi)[0]  # forward
            # cv2.imwrite(f'img_{si}.jpg', 255 * xi[0].cpu().numpy().transpose((1, 2, 0))[:, :, ::-1])  # save
            yi = self._descale_pred(yi, fi, si, img_size)
            y.append(yi)
        return torch.cat(y, 1), None  # augmented inference, train
```
Please correct me if I am wrong, but as I understand it, in this setting the TTA is predicting every image 3 times:

1. Scale 1 without flipping
2. Scale 0.83 flipped upside down (EDIT - left/right!)
3. Scale 0.67 without flipping

After prediction, the inferred bounding boxes are "deaugmented" to the original scale and rotation/flip, concatenated together and then the NMS is performed on all the predicted boxes to filter the final output bounding boxes.

My question is:

1. Do you think implementing rot90 rotation to the TTA pipeline is viable? If yes, do you think it is a good thing for a PR? 
2. As the Yolov5 is already integrated with Albumentations library, would it make sense to integrate for example Qubvels TTA library in Pytorch? https://github.com/qubvel/ttach
3. Is my understanding of current TTA implementation correct - is it an "average" of 3 augmented predictions?

Thanks,

Martin

**Comment by glenn-jocher** [comment]
@mrkolarik yes there's a lot of space for R&D and improvement on the TTA side alone. The current TTA runs 3 images at different scales with a left-right flip on the middle one. This was a good middle ground in our testing. Additional ops help more but of course at added expense.

It would be nice to create a more generalized implementation where you could program a custom pipeline to include ops like your 90 deg rotation. We have powerful built-in augmentations in the dataloader, and recently introduce an Albumentations integration also, but none of these offer easy automatic de-augmentation of labels using the inverse transforms.

Average is not used, you can see in the code you passed the results are concatenated. Can you explain about the qubvel library?

**Comment by mrkolarik** [comment]
Thanks for your quick answer @glenn-jocher ! Regarding your questions:

1. The Qubvel library - I found this when I was looking for solutions how to implement more advanced TTA and this library by Qubvel is in pure Pytorch and hopefully will be fast enough. It was one of my possible solutions on TTA implementation and I wanted to know your opinion, as it includes both augmentation and de-augmentation implementations and is MIT licensed.
2. I used the term "average" here in misleading way to express the somewhat general idea regarding TTA, however I understand the results are concatenated and not averaged

I could have a look at implementing a general TTA pipeline - but as I have never did a PR to your project, it would be nice to consult this with someone who has more knowledge regarding your whole repository. If you think, this would be good and worthwile, could you point me to someone?

**Comment by glenn-jocher** [comment]
@BloodAxe CC'ing you here on this TTA issue (https://github.com/ultralytics/yolov5/issues/4394 similar also). 

Users are requesting specific transforms for TTA (i.e. rot90 here), proposing Qubvel integration, but I think it would be more useful to extend the Albumentations capability to TTA to allow for any combination of transforms. It seems like the major capability that's lacking is inverting the predictions (using the inverse transform). A possible high-level workflow might be:

```python
im2, xform = self.transform(im)  # Albumentations TTA
results2 = model(im2)  # inference
results = xform.inverse(results2)  # invert predictions
```

**Comment by glenn-jocher** [comment]
@mrkolarik I've opened up a feature request for this with albumentations in https://github.com/albumentations-team/albumentations/issues/987
